src/components/ApiKeys.vue

```diff
@@ -0,0 +1,228 @@
+<script setup lang="ts">
+import { ref } from 'vue'
+import { useI18n } from 'vue-i18n'
+import { toast } from 'vue-sonner'
+import copy from 'copy-text-to-clipboard'
+import { useSupabase } from '~/services/supabase'
+import type { Database } from '~/types/supabase.types'
+import { useMainStore } from '~/stores/main'
+import { useDisplayStore } from '~/stores/display'
+import Plus from '~icons/heroicons/plus'
+import Trash from '~icons/heroicons/trash'
+import Pencil from '~icons/heroicons/pencil'
+import Clipboard from '~icons/heroicons/clipboard-document'
+
+const { t } = useI18n()
+const displayStore = useDisplayStore()
+const main = useMainStore()
+const isLoading = ref(false)
+const supabase = useSupabase()
+const apps = ref<Database['public']['Tables']['apikeys']['Row'][]>()
+async function getKeys(retry = true): Promise<void> {
+  isLoading.value = true
+  const { data } = await supabase
+    .from('apikeys')
+    .select()
+    .eq('user_id', main.user?.id)
+  if (data && data.length)
+    apps.value = data
+
+  else if (retry && main.user?.id)
+    return getKeys(false)
+
+  isLoading.value = false
+}
+displayStore.NavTitle = ''
+getKeys()
+
+async function addNewApiKey() {
+  if (await showAddNewKeyModal())
+    return
+
+  const keyType = displayStore.lastButtonRole
+
+  let databaseKeyType: 'read' | 'write' | 'all' | 'upload'
+
+  switch (keyType) {
+    case 'cancel':
+    case '':
+      return
+    case 'read-button':
+      databaseKeyType = 'read'
+      break
+    case 'upload-button':
+      databaseKeyType = 'upload'
+      break
+    case 'all-button':
+      databaseKeyType = 'all'
+      break
+    default:
+      return
+  }
+
+  const newApiKey = crypto.randomUUID()
+  const { data: { user } } = await supabase.auth.getUser()
+
+  if (!user) {
+    console.log('Not logged in, cannot regenerate API key')
+    return
+  }
+
+  const { data, error } = await supabase
+    .from('apikeys')
+    .upsert({ user_id: user.id, key: newApiKey, mode: databaseKeyType })
+    .select()
+
+  if (error)
+    throw error
+
+  apps.value?.push(data[0])
+  toast.success(t('add-api-key'))
+}
+
+async function regenrateKey(app: Database['public']['Tables']['apikeys']['Row']) {
+  if (await showRegenerateKeyModal())
+    return
+
+  const newApiKey = crypto.randomUUID()
+  const { data: { user } } = await supabase.auth.getUser()
+
+  if (!user) {
+    console.log('Not logged in, cannot regenerate API key')
+    return
+  }
+
+  const { error } = await supabase
+    .from('apikeys')
+    .update({ key: newApiKey })
+    .eq('user_id', user.id)
+    .eq('key', app.key)
+
+  if (error || typeof newApiKey !== 'string')
+    throw error
```

Comment:
error could be undefined and newApiKey not a string, this would lead to unexpected behavior right?